src/backend/distributed/commands/database.c

```diff
@@ -0,0 +1,202 @@
+/*-------------------------------------------------------------------------
+ *
+ * database.c
+ *    Commands to interact with the database object in a distributed
+ *    environment.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/htup_details.h"
+#include "access/xact.h"
+#include "catalog/objectaddress.h"
+#include "catalog/pg_database.h"
+#include "commands/dbcommands.h"
+#include "miscadmin.h"
+#include "nodes/parsenodes.h"
+#include "utils/syscache.h"
+
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_executor.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/worker_transaction.h"
+
+static void EnsureSequentialModeForDatabaseDDL(void);
+static AlterOwnerStmt * RecreateAlterDatabaseOwnerStmt(Oid databaseOid);
+static Oid get_database_owner(Oid db_oid);
+
+/* controlled via GUC */
+bool EnableAlterDatabaseOwner = false;
+
+
+/*
+ * PreprocessAlterDatabaseOwnerStmt is called during the utility hook before the alter
+ * command is applied locally on the coordinator. This will verify if the command needs to
+ * be propagated to the workers and if so prepares a list of ddl commands to execute.
+ */
+List *
+PreprocessAlterDatabaseOwnerStmt(Node *node, const char *queryString,
+								 ProcessUtilityContext processUtilityContext)
+{
+	AlterOwnerStmt *stmt = castNode(AlterOwnerStmt, node);
+	Assert(stmt->objectType == OBJECT_DATABASE);
+
+	ObjectAddress typeAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
+	if (!ShouldPropagateObject(&typeAddress))
+	{
+		return NIL;
+	}
+
+	if (!EnableAlterDatabaseOwner)
+	{
+		/* don't propagate if GUC is turned off */
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	QualifyTreeNode((Node *) stmt);
+	const char *sql = DeparseTreeNode((Node *) stmt);
+
+	EnsureSequentialModeForDatabaseDDL();
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+List *
+PostprocessAlterDatabaseOwnerStmt(Node *node, const char *queryString)
+{
+	AlterOwnerStmt *stmt = castNode(AlterOwnerStmt, node);
+	Assert(stmt->objectType == OBJECT_DATABASE);
+
+	ObjectAddress typeAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
+	if (!ShouldPropagateObject(&typeAddress))
+	{
+		return NIL;
+	}
+
+	if (!EnableAlterDatabaseOwner)
+	{
+		/* don't propagate if GUC is turned off */
+		return NIL;
+	}
+
+	EnsureDependenciesExistOnAllNodes(&typeAddress);
```

Comment:
I think having `EnsureDependenciesExistOnAllNodes` is very clear and nice.

Though I have a question that I'm curious about the answer. Can there be any objects that is going to be created via the database dependency?  As far as I can see, pg_depend/pg_shdepend doesn't have any entries with `where classid = 'pg_catalog.pg_database'::regclass::oid;`

For example, when I create a database with an owner, I expected that the database depends on the owner. But, it doesn't?

Reply:
>For example, when I create a database with an owner, I expected that the database depends on the owner. But, it doesn't?

Oh, it does, I missed that. And, so this should generate `CREATE ROLE` statements as well on enterprise even if the role has not been created earlier, right?

```SQL
# select * from pg_shdepend where classid = 'pg_catalog.pg_database'::regclass::oid;
 dbid | classid | objid | objsubid | refclassid | refobjid | deptype 
------+---------+-------+----------+------------+----------+---------
    0 |    1262 | 13720 |        0 |       1260 |    16916 | o
(1 row)
```

Reply:
Yeah, from my experiments the database will depend on the user.

users being a supported object in enterprise will make that the creation of the user gets ordered before the altering of the owner.

This, however, is a good test to add in the enterprise repo.